utils/copyright-conformance.py

```diff
@@ -0,0 +1,226 @@
+COPYRIGHT_LITERAL_PATTERN = r'Copyright \(c\) '
+
+DEBUG = False
+VERBOSE = False
+
+
+def get_parser():
+    parser = argparse.ArgumentParser(description='Validates that all source files given on command line \
+                                            have correct copyright headers, based on git modification dates. \
+                                            Note that this edits the modification date.')
+    required = parser.add_argument_group('required arguments')
+    required.add_argument('-f', nargs='+', help='One or more files/directories to check', required=True)
+    parser.add_argument('-r', '--recursive', help='Recursively search through directories', action="store_true", default=False, required=False)
+    parser.add_argument('-d', '--debug', help="Print files whose copyright headers could not be found, or don't exist", action='store_true', default=False, required=False)
+    parser.add_argument('-v', '--verbose', help="Print additional info on when copyright headers are conformant, or when files are patched", action='store_true', default=False, required=False)
+    parser.add_argument('-p', '--patch', help='Replace non-conforming copyright headers with the correct years', action="store_true", default=False, required=False)
+    parser.add_argument('-i', '--interactive', help='Shows copyright notice before and after change. Asks for user confirmation when patching files', action="store_true", default=False, required=False)
+    return parser
+
+
+def debug(message):
+    global DEBUG
+    if DEBUG:
+        print(message)
+
+
+def verbose(message):
+    global VERBOSE
+    if VERBOSE:
+        print(message)
+
+
+def discover_files(filepaths, recursive_opt):
+    """Make sure all files given exist, and discover any files in folders if we're discovering files recursively."""
+    # TODO: Make paths relative to swift source directory
+    discovered_files = set()
+
+    i = 0
+    while i < len(filepaths):
+        f = os.path.abspath(filepaths[i])
+        # exit if files given on command line don't exist
+        if not os.path.exists(f):
+            print(f"Unexpected error: No such file or directory {f}", file=sys.stderr)
+            sys.exit(1)
+
+        if os.path.isdir(f):
+            if not recursive_opt:
+                print(f"You passed in a directory ({f}), without specifying a recursive search. That doesn't make sense because we can't check a directory for conformance. Either omit the directory, or add the --recursive flag.", file=sys.stderr)
+                sys.exit(1)
+            else:
+                dir_contents = [os.path.join(f, path) for path in os.listdir(f)]
+                filepaths.extend(dir_contents)
+        else:
+            discovered_files.add(f)
+
+        i += 1
+    return list(discovered_files)
+
+
+def has_copyright_notice(candidate_filepath):
+    """Checks if a file has a copyright notice"""
+    with open(candidate_filepath, "r") as f:
+        contents = f.read()
```

Comment:
This line can fail.  We should catch any error that happens here and print out the `candidate_filepath`

Reply:
yup, i see. it looks like some non-unicode files have been added to the tree in the last 4 years

i've pushed a change that skips all files that fail to read with a non-unicode error